### PR TITLE
Added GET and POST /addresses endpoint

### PR DESCRIPTION
**Pull request by begedin**
- [Add endpoint to post location - Phase 1 (without scoring)](https://app.asana.com/0/52390949460913/53022006376852)
- [Add scoring to POST location endpoint - Phase 2](https://app.asana.com/0/52390949460913/53022006376865)
# Description

Adds POST /addresses and GET /addresses endpoint

I'm not sure if that's the endpoint name we want, or if should be "/location", or something else completely.

At the moment, the endpoints still do nothing, but we do have an address model with a `.within` scope which can be used to find a list of addresses within a bounding box. My logic was that the map API we're using in the client app probably has some sort of `getViewBox` method.
### 1. Update an existing address

I'm guessing the user will make the edits to the (partially filled) actual address fields, then that data, alongside the coordinates, will be sent to the API. So something like 

``` Javascript
{  
  longitude, latitude, street_1, street_2, 
  city, state_code, zip_code, result (an enum string)
}
```

There is just one endpoint (`create`), but depending the API should decide if it's actually a new record, or an existing one, based on several of the fields it receives. The `visited_at` field get's updated with the current time stamp in any case.

This part of determining if the address is new or exiting might be a problem, though. We cant rely on coordinates alone, unless we round them down to a point, or decide on some arbitrary minimum distance to consider one location different from another. 

I guess we could go address fields first, and then coordinates if parts of the address are missing. It really depends on what the exact flow will be.
### 2. Get a list of potential addresses

The request sends  the following parameters: `{ latitude, longitude, radius }` and the API returns a list of addresses within the circle specified by those parameters.

### CODE REVIEW

File: app/controllers/addresses_controller.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+  def index
+    render json: index_params
+  end
+
+  def create
+  end
+
+  def index_params
+    lattitude = params.require(:lattitude)
```
Comment (joshsmith) [inline app/controllers/addresses_controller.rb]: This is misspelled. It should be `latitude`.

**Comment by joshsmith** [comment]
Does the client ever need to update addresses? I don't think they would. I think it would be one and done from the client's perspective. We only send when the person is done at the person's doorstep and the coordinates, address, result, etc. get `POST`-ed.

I think this brings a greater question to the fore: do we have an address that stores the actual addresses people have been to, and one for the "true" address information. That is, we have true address information which is a find or create, and then the attempt at the address (the door-knocking effort).

I'm not sure. I need to think on this more. And go to dinner.

As for (2), we discussed sending a lat/long and radius for the `GET` request and sending back our logged addresses. Definitely feel free to move forward on this regardless of the other situation.

**Comment by begedin** [comment]
@JoshSmith I did some testing and considering the results, we may actually be better off using regular postgres with geokit-rails

# postgis with RGeo/active_postgis_adapter

**Scenario**: We create X records, and then use the `:within` scope to find all records within a certain distance.

Results are in seconds. First is creation time, which should not matter much. Second is fetch time.

## 1000 records

2.124975565
0.164099854

## 10000 records

18.079103786
0.205285486

## 20000 records

36.200978929
0.176882954

## 30000 records

58.318695763
0.212598539

# postgres with geokit-rails

## 1000 records

1.755789673
0.032293041

## 10000 records

17.99527405
0.082253743

## 20000 records

36.674598752
0.147831757

### 30000 records

56.647421059
0.131549736

The values are averages and as you can see, with regular postgres and a higher record number, they're really close together, to within a margin of error.

Overall, it's relatively easy to switch from one to another at this point, so I can continue work for now and then make the final changes when we decide which direction to go in.

**Comment by begedin** [comment]
@JoshSmith 

Regarding the client needing to update an address, my imagined scenario was a client visiting a residence that was previously marked as "not_home", or potentially even "not_interested" and updating it with something else. Is that scenario not supported?

**Comment by begedin** [comment]
Regarding CI failing, my guess is, since I changed an existing migration, it isn't passing. Can it be configured so it does a db reset, or do I do a separate migration to switch between postgis and regular postgres?

While we're at it, there's a question about the `visited_at` field. 
- [ ] Is there any specific case where it should or shouldn't update to the current time? If not, we could just use timestamps provided by active record.

**Comment by joshsmith** [comment]
I think the scenario is supported for visiting a residence previously marked as `not_home`, but I think this calls for another table, with maybe a caching of `latest_result` in this table for easier fetching of the results set without joins. I'd then add in a `last_visited_at` timestamp.

We should be tracking every attempt and by whom, i.e. a `users_addresses`, that includes their specific results. If 5 different people knock on the same door, we should have that information.

**Comment by joshsmith** [comment]
Really awesome work on the performance testing. Exactly what we needed. I think it's obvious we should use `geokit-rails` for now, especially since you said the change is not hard to make down the road if we need to.

How were you able to generate the sample records? Wondering if it's something I could use for testing map display locally.

**Pull request merged**